### PR TITLE
ddl: add partion function check when create partion table

### PR DESCRIPTION
**Pull request by crazycs520**
## What have you changed? (mandatory)
* check partition function valid when create partition table.
Fix issue[#7058](https://github.com/pingcap/tidb/issues/7058)

MySQL Partitioning Limitations Relating to Functions:
https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.7/en/partitioning-limitations-functions.html

## What is the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

Unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No
## Does this PR affect tidb-ansible update? (mandatory)

No

## Does this PR need to be added to the release notes? (mandatory)
No

### CODE REVIEW

**Comment by crazycs520** [comment]
@tiancaiamao @ciscoxll @winkyao @zimulala

**Comment by winkyao** [comment]
@crazycs520 Well done, LGTM

File: ddl/partition.go [hunk 1]
```diff
@@ -112,6 +112,26 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
 	return nil
 }
 
+func checkPartitionFuncValid(expr ast.ExprNode) error {
+	switch v := expr.(type) {
+	case *ast.CaseExpr:
+		return ErrPartitionFunctionIsNotAllowed
+	case *ast.FuncCallExpr:
+		// check function which allowed in partitioning expressions
+		// https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.5/en/partitioning-limitations-functions.html
```
Comment (shenli) [inline ddl/partition.go]: Please use the MySQL-5.7's document.

File: ddl/partition.go [hunk 2]
```diff
@@ -112,6 +112,26 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
 	return nil
 }
 
+func checkPartitionFuncValid(expr ast.ExprNode) error {
```
Comment (shenli) [inline ddl/partition.go]: The expression should return integer value:

> Data type of partitioning key.  A partitioning key must be either an integer column or an expression that resolves to an integer. Expressions employing ENUM columns cannot be used. The column or expression value may also be NULL. (See Section 3.7, “How MySQL Partitioning Handles NULL”.)
https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.7/en/partitioning-limitations.html

File: ddl/partition.go [hunk 3]
```diff
@@ -112,6 +112,26 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+	switch v := expr.(type) {
+	case *ast.CaseExpr:
```
Comment (ciscoxll) [inline ddl/partition.go]: The expression with ` + - * ` arithmetic can inside.

File: ddl/partition.go [hunk 4]
```diff
@@ -112,6 +112,26 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+		return ErrPartitionFunctionIsNotAllowed
+	case *ast.FuncCallExpr:
+		// check function which allowed in partitioning expressions
+		// https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.5/en/partitioning-limitations-functions.html
+		switch v.FnName.L {
```
Comment (tiancaiamao) [inline ddl/partition.go]: And the test code should cover more branches

File: ddl/partition.go (hunk 3 above)
Comment (crazycs520) [reply ddl/partition.go]: done.PTAL

File: ddl/partition.go [hunk 5]
```diff
@@ -112,6 +113,33 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
 	return nil
 }
 
+func checkPartitionFuncValid(expr ast.ExprNode) error {
+	switch v := expr.(type) {
+	case *ast.CaseExpr:
+		return ErrPartitionFunctionIsNotAllowed
+	case *ast.FuncCallExpr:
+		// check function which allowed in partitioning expressions
+		// https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.7/en/partitioning-limitations-functions.html
```
Comment (zimulala) [inline ddl/partition.go]: s/` https`/`See https`

File: ddl/partition.go [hunk 6]
```diff
@@ -112,6 +113,33 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+		case ast.Abs, ast.Ceiling, ast.DateDiff, ast.Day, ast.DayOfMonth, ast.DayOfWeek, ast.DayOfYear, ast.Extract, ast.Floor,
+			ast.Hour, ast.MicroSecond, ast.Minute, ast.Mod, ast.Month, ast.Quarter, ast.Second, ast.TimeToSec, ast.ToDays,
+			ast.ToSeconds, ast.UnixTimestamp, ast.Weekday, ast.Year, ast.YearWeek:
+			return nil
+		default:
+			return ErrPartitionFunctionIsNotAllowed
+		}
+	case *ast.BinaryOperationExpr:
+		// The DIV operator is also supported; the / operator is not permitted.
+		// https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations.html
```
Comment (zimulala) [inline ddl/partition.go]: Ditto.

File: ddl/partition.go [hunk 7]
```diff
@@ -112,6 +113,33 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+		if v.Op == opcode.Div {
+			return ErrPartitionFunctionIsNotAllowed
+		}
+		return nil
+	default:
```
Comment (zimulala) [inline ddl/partition.go]: Cloud we remove this line.

Comment (zimulala) [inline ddl/partition.go]: Cloud we remove this two lines.

File: ddl/db_test.go [hunk 8]
```diff
@@ -1666,14 +1666,24 @@ func (s *testDBSuite) TestCreateTableWithPartition(c *C) {
...
 	partition by range( case when col1 > 0 then 10 else 20 end ) (
 		partition p0 values less than (2),
 		partition p1 values less than (6)
 	);`)
+	c.Assert(ddl.ErrPartitionFunctionIsNotAllowed.Equal(err), IsTrue)
+
+	_, err = s.tk.Exec(`create TABLE t10 (c1 int,c2 int) partition by range(c1 / c2 ) (partition p0 values less than (2));`)
+	c.Assert(ddl.ErrPartitionFunctionIsNotAllowed.Equal(err), IsTrue)
+	_, err = s.tk.Exec(`create TABLE t11 (c1 int,c2 int) partition by range(c1 div c2 ) (partition p0 values less than (2));`)
+	c.Assert(err, IsNil)
```
Comment (zimulala) [inline ddl/db_test.go]: s/`Exec`/`MustExec`
Then we can remove `c.Assert(err, IsNil)`.

File: ddl/db_test.go [hunk 9]
```diff
@@ -1666,14 +1666,24 @@ func (s *testDBSuite) TestCreateTableWithPartition(c *C) {
...
 	col1 int
 	)
 	partition by range( case when col1 > 0 then 10 else 20 end ) (
 		partition p0 values less than (2),
 		partition p1 values less than (6)
 	);`)
+	c.Assert(ddl.ErrPartitionFunctionIsNotAllowed.Equal(err), IsTrue)
+
+	_, err = s.tk.Exec(`create TABLE t10 (c1 int,c2 int) partition by range(c1 / c2 ) (partition p0 values less than (2));`)
+	c.Assert(ddl.ErrPartitionFunctionIsNotAllowed.Equal(err), IsTrue)
```
Comment (zimulala) [inline ddl/db_test.go]: Could we use the `testErrorCode` directly?

File: ddl/partition.go [hunk 10]
```diff
@@ -112,6 +115,51 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+		return nil
+	}
+}
+
+func checkPartitionFuncType(ctx sessionctx.Context, expr ast.ExprNode, tblInfo *model.TableInfo) error {
```
Comment (XuHuaiyu) [inline ddl/partition.go]: add a comment for this functions.

File: ddl/partition.go [hunk 11]
```diff
@@ -112,6 +115,51 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
...
+			ast.Hour, ast.MicroSecond, ast.Minute, ast.Mod, ast.Month, ast.Quarter, ast.Second, ast.TimeToSec, ast.ToDays,
+			ast.ToSeconds, ast.UnixTimestamp, ast.Weekday, ast.Year, ast.YearWeek:
+			return nil
+		default:
+			return ErrPartitionFunctionIsNotAllowed
+		}
+	case *ast.BinaryOperationExpr:
+		// The DIV operator is also supported; the / operator is not permitted.
+		// https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations.html
+		if v.Op == opcode.Div {
```
Comment (XuHuaiyu) [inline ddl/partition.go]: since DIV is supported why we return error here?

File: ddl/partition.go [hunk 12]
```diff
@@ -112,6 +115,51 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
 	return nil
 }
 
+func checkPartitionFuncValid(expr ast.ExprNode) error {
```
Comment (XuHuaiyu) [inline ddl/partition.go]: add a comment for this function.

File: ddl/ddl.go [hunk 13]
```diff
@@ -178,6 +178,10 @@ var (
 	ErrDropLastPartition = terror.ClassDDL.New(codeDropLastPartition, mysql.MySQLErrName[mysql.ErrDropLastPartition])
 	//ErrTooManyPartitions returns too many partitions were defined.
 	ErrTooManyPartitions = terror.ClassDDL.New(codeTooManyPartitions, mysql.MySQLErrName[mysql.ErrTooManyPartitions])
+	//ErrPartitionFunctionIsNotAllowed returns this partition function is not allowed.
+	ErrPartitionFunctionIsNotAllowed = terror.ClassDDL.New(codePartitionFunctionIsNotAllowed, mysql.MySQLErrName[mysql.ErrPartitionFunctionIsNotAllowed])
+	//
```
Comment (XuHuaiyu) [inline ddl/ddl.go]: miss comment

File: ddl/partition.go (hunk 11 above)
Comment (crazycs520) [reply ddl/partition.go]: DIV operator is `opcode.IntDiv`, / operator is `opcode.Div`

File: ddl/partition.go (hunk 2 above)
Comment (crazycs520) [reply ddl/partition.go]: done

**Comment by tiancaiamao** [comment]
LGTM @winkyao @zimulala

File: ddl/ddl_api.go [hunk 14]
```diff
@@ -849,6 +849,14 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 		if err != nil {
 			return errors.Trace(err)
 		}
+		err = checkPartitionFuncValid(s.Partition.Expr)
```
Comment (ciscoxll) [inline ddl/ddl_api.go]: Please move the code to `partition.go` 52 line.

Comment (ciscoxll) [reply ddl/ddl_api.go]: LGTM

**Comment by tiancaiamao** [comment]
It would nice if maybe you could resolve the conflicts. @crazycs520

**Comment by crazycs520** [comment]
I have reorder the check sequence when create partition table.
```SQL
create TABLE t25 (c1 float) partition by range( c1 ) (
	partition p0 values less than maxvalue,
	partition p1 values less than (2000)
);
```
The SQL have 2 problem:
1. partition by a float column.		error is: `ErrFieldTypeNotAllowedAsPartitionField`
2. `maxvalue` should be the last line. error is: `ErrPartitionMaxvalue`

mysql execute the sql will return is: `ErrPartitionMaxvalue`
TiDB  execute the sql use to return `ErrFieldTypeNotAllowedAsPartitionField`

So I reorder the check sequence.
@ciscoxll  @tiancaiamao PTAL

**Comment by ciscoxll** [comment]
Again to determine LGTM.

**Pull request merged**